From: Niko Tyni <ntyni@debian.org>
Subject: Fix File::Copy::copy with pipes on GNU/kFreeBSD

Quoting Petr Salinger in http://bugs.debian.org/537555:
The Copy tries to detect whether source and dest are the same files.
Unfortunately, on the GNU/kFreeBSD the kernel returns for all pipes
as device and inode numbers just zero. See pipe_stat() in
http://www.freebsd.org/cgi/cvsweb.cgi/src/sys/kern/sys_pipe.c

Patch by Petr Salinger, tests by Niko Tyni.

Backported from blead change 16f708c9bc0dc48713b200031295a40bed83bbfc

---
 lib/File/Copy.pm |    2 +-
 lib/File/Copy.t  |   15 ++++++++++++++-
 2 files changed, 15 insertions(+), 2 deletions(-)

diff --git a/lib/File/Copy.pm b/lib/File/Copy.pm
index 1520e0a..c7d5ded 100644
--- a/lib/File/Copy.pm
+++ b/lib/File/Copy.pm
@@ -102,7 +102,7 @@ sub copy {
 	my @fs = stat($from);
 	if (@fs) {
 	    my @ts = stat($to);
-	    if (@ts && $fs[0] == $ts[0] && $fs[1] == $ts[1]) {
+	    if (@ts && $fs[0] == $ts[0] && $fs[1] == $ts[1] && !-p $from) {
 		carp("'$from' and '$to' are identical (not copied)");
                 return 0;
 	    }
diff --git a/lib/File/Copy.t b/lib/File/Copy.t
index 84abfd5..294aa55 100755
--- a/lib/File/Copy.t
+++ b/lib/File/Copy.t
@@ -11,7 +11,7 @@ use Test::More;
 
 my $TB = Test::More->builder;
 
-plan tests => 60;
+plan tests => 62;
 
 # We're going to override rename() later on but Perl has to see an override
 # at compile time to honor it.
@@ -191,6 +191,19 @@ for my $cross_partition_test (0..1) {
 }
 
 
+SKIP: {
+    skip("fork required to test pipe copying", 2)
+        if (!$Config{'d_fork'});
+
+    open(my $IN, "-|") || exec $^X, '-e', 'print "Hello, world!\n"';
+    open(my $OUT, "|-") || exec $^X, '-ne', 'exit(/Hello/ ? 55 : 0)';
+
+    ok(copy($IN, $OUT), "copy pipe to another");
+    close($OUT);
+    is($? >> 8, 55, "content copied through the pipes");
+    close($IN);
+}
+
 END {
     1 while unlink "file-$$";
     1 while unlink "lib/file-$$";
-- 
tg: (71b2123..) fixes/kfreebsd-filecopy-pipes (depends on: upstream)
